src/posix/net.cpp

```diff
@@ -426,42 +465,143 @@ void CurlHttpIO::closecurlevents(direction_t d)
             WSACloseEvent(info.handle);
         }
     }
+#endif
     socketmap.clear();
 }
 
+void CurlHttpIO::processaresevents()
+{
+#if !(defined(_WIN32) && !defined(WINDOWS_PHONE))
+    fd_set *rfds = &((PosixWaiter *)waiter)->rfds;
+    fd_set *wfds = &((PosixWaiter *)waiter)->wfds;
+#endif
+
+    for (unsigned int i = 0; i < aressockets.size(); i++)
+    {
+        SockInfo &info = aressockets[i];
+        if (!info.mode)
+        {
+            continue;
+        }
+
+#if defined(_WIN32) && !defined(WINDOWS_PHONE)
+        if (info.handle == WSA_INVALID_EVENT)
+        {
+            continue;
+        }
+
+        if (WSAWaitForMultipleEvents(1, &info.handle, TRUE, 0, FALSE) == WSA_WAIT_EVENT_0)
+        {
+            WSAResetEvent(info.handle);
+            ares_process_fd(ares,
+                            (info.mode & SockInfo::READ) ? info.fd : ARES_SOCKET_BAD,
+                            (info.mode & SockInfo::WRITE) ? info.fd : ARES_SOCKET_BAD);
+        }
+#else
+        if (((info.mode & SockInfo::READ) && FD_ISSET(info.fd, rfds)) || ((info.mode & SockInfo::WRITE) && FD_ISSET(info.fd, wfds)))
+        {
+            ares_process_fd(ares,
+                            ((info.mode & SockInfo::READ) && FD_ISSET(info.fd, rfds)) ? info.fd : ARES_SOCKET_BAD,
+                            ((info.mode & SockInfo::WRITE) && FD_ISSET(info.fd, wfds)) ? info.fd : ARES_SOCKET_BAD);
+        }
+#endif
+    }
+
+    if (arestimeout >= 0 && arestimeout <= Waiter::ds)
+    {
+        arestimeout = -1;
+        ares_process_fd(ares, ARES_SOCKET_BAD, ARES_SOCKET_BAD);
+    }
+}
+
 void CurlHttpIO::processcurlevents(direction_t d)
 {
+#if !(defined(_WIN32) && !defined(WINDOWS_PHONE))
+    fd_set *rfds = &((PosixWaiter *)waiter)->rfds;
+    fd_set *wfds = &((PosixWaiter *)waiter)->wfds;
+#endif
+
     int dummy = 0;
-    std::map<int, SockInfo> &socketmap = (d == API) ? curlapisockets : ((d == GET) ? curldownloadsockets : curluploadsockets);
+    bool isapipaused = false;
+    std::map<int, SockInfo> *socketmap;
+    m_time_t *timeout;
+    bool *paused;
 
-    bool active = false;
-    for (std::map<int, SockInfo>::iterator it = socketmap.begin(); it != socketmap.end();)
+    if (d == API)
```

Comment:
If we had declared these variables like:
```
curlsockets[3];
```
and so on, we could use the `d` as index to access to the corresponding curl socket, curl timeout reset, number of connections...